### PR TITLE
doc: actually document `lib.customisation.makeScope`

### PR DESCRIPTION
**Pull request by fricklerhandwerk**
## Description of changes

Added interface documentation to `makeScope` that hopefully reduces confusion rather than creating more. Took some parts from https://github.com/NixOS/nixpkgs/pull/270696 by @9999years.

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: lib/customisation.nix [hunk 1]
```diff
@@ -306,18 +306,117 @@ rec {
...
+      bar = «derivation»;
+    }
+    ```
+    :::
+
+    # Inputs
+
+    1. `newScope` (`AttrSet -> ((AttrSet -> a) | Path) -> AttrSet -> a`)
+
+       A function that Takes an attribute set `attrs` and returns what ends up as `callPackage` in the output.
```
Comment (fricklerhandwerk) [inline lib/customisation.nix:345]:
```suggestion
       A function that takes an attribute set `attrs` and returns what ends up as `callPackage` in the output.
```

**Comment by szlend** [comment]
Not sure if this is the right place and time to include this, but something that's caused a lot of confusion for me was that `makeScope` doesn't do any of the splicing shenanigans. So when doing cross-compilation, you'll get incorrect variants of packages in `nativeBuildInputs`.

Is it worth leaving a note and pointing people at the undocumented `makeScopeWithSplicing'`? Or wait until that's documented?

**Comment by fricklerhandwerk** [comment]
True, and I'd rather avoid linking to something that is just jarring. That's in fact the reason I started this piece, because I'm in the process of migrating "the" callPackage tutorial to nix.dev and found reference documentation shockingly lacking. https://github.com/NixOS/nix.dev/pull/935

I'd drop a reference, but honestly I wouldn't know what to even write there, because for me splicing is effectively an undefined term at this point. I'd fix all of that, but it would take me heaps of time I don't have to get into the subject.

It seems not wrong not to mention things that are not explored well-enough to make sense. If people need it enough, the code will always represent itself correctly, while misleading documentation will just add a layer of confusion.

File: lib/customisation.nix [hunk 2]
```diff
@@ -306,18 +306,117 @@ rec {
...
+    {
+      callPackage = «lambda»;
+      newScope = «lambda»;
+      overrideScope = «lambda»;
+      packages = «lambda»;
+      foo = «derivation»;
+      bar = «derivation»;
+    }
+    ```
+    :::
```
Comment (DanielSidhion) [inline lib/customisation.nix:339]: Can you move this example to a `# Examples` header at the end of the docstring, and make a reference to the example here instead? Something like ``See [](#ex-makeScope) for how to use `makeScope` at a high level``.

It helps keep the reading flow uninterrupted while bringing awareness that there are also examples to look at.

File: lib/customisation.nix [hunk 3]
```diff
@@ -306,18 +306,117 @@ rec {
...
-     with the provided `callPackage` will be evaluated with the same
-     arguments. Any package in the set may depend on any other. The
-     `overrideScope'` function allows subsequent modification of the package
-     set in a consistent way, i.e. all packages in the set will be
-     called with the overridden packages. The package sets may be
-     hierarchical: the packages in the set are called with the scope
-     provided by `newScope` and the set provides a `newScope` attribute
-     which can form the parent scope for later package sets.
+  /**
+    Make an attribute set (a "scope") from functions that take arguments from that same attribute set.
```
Comment (DanielSidhion) [inline lib/customisation.nix:310]: I think you did a great job documenting "what" `makeScope` does (it's hard to follow the explanations sometimes due to the high amount of names/concepts involved and some back and forth is needed, but I think that's more due to the complexity of the problem).

I haven't read too much code that uses `makeScope` yet, but I think this documentation is missing a "why would I use this". For example, which use cases/problems does this help me with?

This is more up to you: if you think the audience that comes here will already know _why_ they need something like this, feel free to ignore this. Maybe some other place in the docs or some other different resource would be a better medium to explain why someone would need this. But if that's not the case, I think just mentioning the problems or use cases that this helps with would be already good content.

Comment (fricklerhandwerk) [reply lib/customisation.nix:310]: TBH I don't exactly know what it's even for, and it seems like it's just a convenience wrapper around `callPackageWith` to be able to extend those base attribute sets more easily. The only reason I'm documenting this is because I care about `callPackage`. `callPackageWith` is the actual workhorse that sets up the `override` attribute and deals with the mess from `mkDerivation`. And even that is not the full story, because somewhere in `stdenv` we define the cross-compilation splicing.

So before anyone writes something stupid, I'm in favor of just showing *what* it does, because I can barely imagine anyone using it for non-magical reasons or without knowing what exactly they're doing.

File: lib/customisation.nix [hunk 4]
```diff
@@ -306,18 +306,117 @@ rec {
...
 
-     Type:
-       makeScope :: (AttrSet -> ((AttrSet -> a) | Path) -> AttrSet -> a) -> (AttrSet -> AttrSet) -> AttrSet
+    :::{#ex-makeScope .example}
+    # Create an interdependent package set based on the scope of `pkgs`
```
Comment (lf-) [inline lib/customisation.nix:313]: I don't understand what "the scope of `pkgs`" means here. Does it mean something like "will look in `pkgs` second"?

File: lib/customisation.nix [hunk 5]
```diff
@@ -306,18 +306,117 @@ rec {
...
+    ```
+    :::
+
+    # Inputs
+
+    1. `newScope` (`AttrSet -> ((AttrSet -> a) | Path) -> AttrSet -> a`)
+
+       A function that takes an attribute set `attrs` and returns what ends up as `callPackage` in the output.
+
+       Typical values are `callPackageWith` or `(makeScope callPackageWith f).newScope`.
```
Comment (lf-) [inline lib/customisation.nix:347]: also `pkgs.newScope`, yeah?

File: lib/customisation.nix [hunk 6]
```diff
@@ -306,18 +306,117 @@ rec {
...
+
+    2. `f` (`AttrSet -> AttrSet`)
+
+       A function that takes an attribute set as returned by `makeScope newScope f` (a "scope") and returns any attribute set.
```
Comment (lf-) [inline lib/customisation.nix:351]: Maybe should say that the argument here is the "self" of the attribute set.

File: lib/customisation.nix (hunk 4 above)
Comment (fricklerhandwerk) [reply lib/customisation.nix:313]: Vaguely, yes, technically: all the attributes from the thing returned by `makeScope somethingsomething pkgs` will be available in `self`, such as `callPackage`

File: lib/customisation.nix (hunk 5 above)
Comment (fricklerhandwerk) [reply lib/customisation.nix:347]: Which is a special case of the second example, and only applies to the top level output of Nixpkgs itself. I try to keep the documentation agnostic to specific implementations. I think we had that discussion over at the Nix manual :) My argument to do it that way is: all of these are pure library functions, they could be used anywhere. So documentation should show how to compose those, not how a particular project happens to deeply embed them, even if it happens to be the parent project.

Comment (fricklerhandwerk) [reply lib/customisation.nix]: Reworded it slightly anyway, please check.

**Review by DanielSidhion — APPROVED** [review]

File: lib/customisation.nix [hunk 7]
```diff
@@ -306,18 +306,127 @@ rec {
...
+
+    # Output
+
+    `makeScope` returns an attribute set of a form called `scope`, which also contains the final attributes produced by `f`:
+
+    ```
+    scope :: {
+      callPackage :: ((AttrSet -> a) | Path) -> AttrSet -> a
+      newScope = AttrSet -> scope
+      overrideScope = (AttrSet -> AttrSet) -> scope
```
Comment (lf-) [inline lib/customisation.nix:339]: This type is incorrect, it can also take a self: super: ... overlay, I'm guessing someone detects if it's a function at some stage. As far as I can tell it also does accept the form specified here but I've not sat down at the computer and checked. 

example: https://github.com/NixOS/nixpkgs/blob/8b8bc2c64b2bafa9d1984b89fa4a2268cecf52e5/pkgs/development/compilers/ghcjs/8.10/common-overrides.nix#L15

File: lib/customisation.nix [hunk 8]
```diff
@@ -306,18 +306,127 @@ rec {
...
+
+      See [](#ex-makeScope-callPackage) for how to use it.
+
+    - `newScope` (`AttrSet -> scope`)
+
+      Takes an attribute set `attrs` and returns a scope that extends the original scope.
+
+    - `overrideScope` (`(AttrSet -> AttrSet) -> scope`)
+
+      Takes a function `g` like the argument `f` to `makeScope`, and returns a new scope where values returned by the original `f` are overriddne with values returned by `g`.
```
Comment (lf-) [inline lib/customisation.nix:362]: typo: overridne

**Review by lf- — COMMENTED** [review]
After addressing the errors mentioned, consider this approved by me. 

Thank you so much for writing these docs. They're such an important piece of documentation to have.

File: lib/customisation.nix (hunk 7 above)
Comment (fricklerhandwerk) [reply lib/customisation.nix:339]: Oh thanks, good catch!

File: lib/customisation.nix [hunk 9]
```diff
@@ -357,10 +357,12 @@ rec {
 
       Takes an attribute set `attrs` and returns a scope that extends the original scope.
 
-    - `overrideScope` (`(AttrSet -> AttrSet) -> scope`)
+    - `overrideScope` (`(scope -> scope -> AttrSet) -> scope`)
```
Comment (lf-) [inline lib/customisation.nix:360]: I don't think it *only* accepts the overlay form though!

Comment (fricklerhandwerk) [reply lib/customisation.nix:360]: It does. This is how `extends g f` works.

**Comment by tomberek** [comment]
Thank you!
